---
title: "MASU Paper"
author: "William Hemstrom"
date: "11/21/2021"
output:
  word_document:
    reference_docx: "style_ref.docx"
  pdf_document:
    keep_tex: true
    fig_caption: true
    latex_engine: pdflatex
    fig_width: 7
    fig_height: 6
bibliography: "Citations.bib"
csl: "council-of-science-editors-author-date.csl"
header-includes:
  - \usepackage[left]{lineno}
  - \linenumbers
  - \usepackage{setspace}\doublespacing
  - \usepackage{sectsty}\sectionfont{\fontsize{12}{12}\selectfont}
  - \usepackage{sectsty}\subsectionfont{\normalfont\itshape\fontsize{12}{12}\selectfont}
  - \usepackage[round]{natbib}
indent: true
sansfont: Times New Roman
fontsize: 12pt
---
```{r, echo=FALSE, include=FALSE}
dat<-read.table("../data/genotypes.geno")                        
meta<- read.table("../data/sample_meta.txt", header=TRUE)
snp.meta<- dat[,1:2]
colnames(snp.meta)<- c("chr", "position")
library(snpR)
dat<- import.snpR.data(dat[,-c(1:2)], snp.meta = snp.meta, sample.meta = meta)
dat<- filter_snps(dat, maf = 0.05, hwe = 0.000001)
number_snps_after_filtering <- nsnps(dat)

#calculating genetic diversity
dat <- calc_ho(dat, c("pop", "phen", "pop.phen"))
get.snpR.stats(dat, c("phen", "pop","pop.phen"), stats="ho") $weighted.means
get.snpR.stats(dat, c("pop", "phen", "pop.phen"))
dat <- calc_pi(dat, c("pop"))
get.snpR.stats(dat, c("pop"), stats="pi") $weighted.means
get.snpR.stats(dat, "pop")
dat <- calc_het_hom_ratio(dat, c("pop", "phen", "pop.phen"))
get.snpR.stats(dat, c("pop", "phen", "pop.phen"))
dat <- calc_maf(dat, c("pop"))
get.snpR.stats(dat, c("pop"))
p <- plot_clusters(dat, facets = c("pop.phen"), plot_type = "umap")
dat <- calc_private(dat, c("phen", "pop","pop.phen"))
get.snpR.stats(dat, c("pop"), stats="private") $weighted.means
get.snpR.stats(dat, c("pop", "phen", "pop.phen"))
dat <- calc_pairwise_fst(dat, c("phen", "pop","pop.phen"), boot = 100, boot_par = 2)
get.snpR.stats(dat, c("phen", "pop","pop.phen"), stats="fst") $weighted.means
fts <- dat[[2]]
dat <- dat[[1]]
stats <- get.snpR.stats(dat, "pop.phen", type = "pairwise")
head(stats)

gp<- run_genomic_prediction(dat, response = "phen", iterations = 10000, burn_in = 1000, thin = 100)
library(data.table)
cv<- cross_validate_genomic_prediction(dat, response = "phen", iterations = 10000, burn_in = 1000, thin = 100)

saveRDS(dat, "processed_genotypes.RDS")
saveRDS(gp, "gp.RDS")

```
# Methods:
## Genotyping 
We extracted DNA from our samples using the magnetic bead approach [@Ali2016]. We then prepared Restriction Associated Digest (RAD) libraries using the Pst1 restriction enzyme [@Ali2016] and sequenced the results on an Illumina Hi-Seq 4000 using 150bp paired-end sequencing. We aligned the data to the O. mykiss v6 reference genome [@Pearse2019] using Burrows-Wheeler Alignment algorithm Aligner [@Li2009]. We then filtered the data for PCR duplicates and improper pairs using SAMtools [@Li2009]. Next, we called the genotypes by using the SAMtools genotype likelihood models [@Li2009] used in the ANGSD software package with a minimum mapping and base call quality score of 20, a SNP p-value of 1e-8, and a minimum acceptable minor allele frequency of 0.05 [@Korneliussen2014]. We also used a minor allele frequency filter (MAF) and a Hardy-Weinberg equilibrium filter using snpR when calling the genotypes [@Wigginton2005]. 
<!-- flagstat google (data script i wrote) -->

```{r}
mat <- read.table("../angsdput.covMat")
IBS <- data.table::fread("../angsdput.covMat")
nrow(IBS)
IBS <- IBS[-32,-c(97, 32)]
pca_r <- eigen(IBS)
pca <- pca_r$vectors
write.table(pca[,1:20], "pca20.txt", col.names = FALSE, row.names = FALSE, quote=FALSE, sep="\t")
colnames(pca) <- paste0("PC", 1:ncol(pca))
meta <- read.table("../data/sample_meta.txt", header = TRUE)[-32,]
pca <- cbind(pca, meta)

loadings <- pca_r$values/sum(pca_r$values)
loadings <- round(loadings*100, 2)
pca$pop <- as.factor(pca$pop)
pca$phen <- as.factor(pca$phen)

library(ggplot2)
p1 <- ggplot(pca, aes(x = PC1, y = PC2, color = pop, shape = phen)) + theme_bw() +
  scale_color_viridis_d() + geom_point(size = 4) + 
  ggplot2::xlab(paste0("PC1 (", loadings[1], "%)")) + 
  ggplot2::ylab(paste0("PC2 (", loadings[2], "%)"))
p1
pca.centroids <- aggregate(pca[,1:2], list(stream = pca$pop), mean)
combine <- rbind(pca[, 1:2], pca.centroids[,-1])
dcombine<- dist(combine, upper = TRUE) 
dcombine<- as.matrix(dcombine)
dcombine<-dcombine[(nrow(dcombine) - nrow(pca.centroids) + 1):nrow(dcombine), -((nrow(dcombine) - nrow(pca.centroids) + 1):nrow(dcombine))]
dcombine <- t(dcombine) 
dcombine <- as.data.frame(dcombine)
dcombine$pop <- pca$pop
dcombine$phen <- pca$phen
colnames(dcombine)[1:2] <- c("CH", "TK")
# ggplot(dcombine, aes(x=pop, y=d, color=phen)) + geom_boxplot() + theme_bw()
dcombine$phen <- factor(dcombine$phen)
dcombine$pop <- factor(dcombine$pop)
mod <- aov(TK~pop+phen+pop*phen, dcombine)
dcombine$d <- 0
dcombine$d[dcombine$pop=="CH"] <- dcombine$TK[dcombine$pop=="CH"]
dcombine$d[dcombine$pop=="TK"] <- dcombine$CH[dcombine$pop=="TK"]
mod <- aov(d~phen+pop+pop*phen, dcombine[-32,])
mod2 <- aov(d~phen, dcombine[-32,])
anova(mod, mod2)
ggplot(dcombine, aes(x=pop, y=d, color=phen)) + geom_boxplot() + theme_bw() + scale_color_viridis_d(end = 0.5)


# library(ggbiplot)
# wine <- iris [, 1:4]
# wine.pca <- prcomp(wine, center = TRUE, scale. = TRUE)
# df.wine.x <- as.data.frame(wine.pca$x)
# df.wine.x$groups <- iris$Species
# pca.centroids <- aggregate(df.wine.x[,1:4], list(groups = df.wine.x$groups), mean)
# library(ggplot2)
# ggplot(df.wine.x, aes(x=PC1, y=PC2, color=groups)) + geom_point() + theme_bw() + 
#   geom_point(data=pca.centroids, color="red", size=10, alpha=0.5) + facet_wrap(~ groups)
# dist(df.wine.x [,-1])
# combine <- rbind(df.wine.x[,-ncol(df.wine.x)], pca.centroids[,-1])
# dcombine<- dist(combine, upper = TRUE) 
```

##  Diversity Stats
We calculated the observed heterozygosity (Ho), heterozygote/homozygote ratio per individual (Het/Hom), the average number of pairwise differences (pi), and fixation index (Fst) using the snpR R package [@Hohenlohe2010; @Wier1984; @Hemstrom2021]. 
<!-- use table to report diversity stats in results (which ones most diverse) get .snpr stats (in uppoer script)-->
<!-- want ngsadmix plot, manhattan plot, pca chart with centroids (two panel one side pca and box and whiskers plots) -->

```{r}
dat <- run_random_forest(dat, response = "phen", mtry = floor(nrow(dat)/2), num.trees = 100000)
RF <- dat$models
dat <- dat$data
summary(RF$.base_.base)
```

## Association Testing
We performed a basic association test on the data using gmmat to correct for the related among individuals and population structure (). We then performed a random forest analysis in R-studio using the ranger R package [@Wright2017]. We used 100,000 trees and used mtry of half the number of snp's. We used a random forest to predict whether the samples would be resident or migratory. Using the BGLR program, we did a genomic prediction using BayesB distribution to estimate the heritability [@Meuwissen2001]. We used 1000 burn- ins and iterations=10,000. The heritability test was done to determine the degree in which phenotypic variation is due to genetic differences among the salmon populations. 


## Population Structure and Migratory Relatedness
We then used NGSadmix (not sure how to describe this). With our data, we also made a Principle Component Analysis chart (PCA) using the ANGSD-IBS method [@Korneliussen2014]. In the PCA, we included centroids of each cluster. 
<!-- Will need to fill in NGSadmix -->

# Results 
```{r}
MASU_data_table <- readr::read_delim("~/MASU_data_table", 
    delim = "\t", escape_double = FALSE, 
    trim_ws = TRUE, skip = 4)
MASU_data_table <- as.data.frame(MASU_data_table)
MASU_data_table[,3] <- gsub("%", "", MASU_data_table[,3])
MASU_data_table[,5] <- gsub("%", "", MASU_data_table[,5])
MASU_data_table[,3]<- as.numeric(MASU_data_table[,3])
MASU_data_table[,5]<- as.numeric(MASU_data_table[,5])
mean(MASU_data_table[,3], na.rm = TRUE)
mean(MASU_data_table[,5], na.rm = TRUE)
```

##Genotyping 
<!-- using linux raw count of # of reads -> how many reads we actually mapped (/home/hemstrow/moths/bams/W/run_read_count.sh) bamlists that are not filtered(.bam) and .sort.flt and import into R and take sum of columns that has the read count run the script twice-->  

unfiltered reads = 163200129
```{r setup, include=FALSE, eval=FALSE}
library(readr)
sort_read_count <- read.csv("~/Downloads/sort.read_count.txt", header = FALSE)
View(sort_read_count)
sort_read_count <- sum(sort_read_count$V1)

sort_flt_read_count <- read.csv("~/Downloads/sort.flt.read_count.txt", header = FALSE)
View(sort_flt_read_count)
sort_flt_read_count <- sum(sort_flt_read_count$V1)

```

flt read = `r sort_flt_read_count` 
<!-- flt read = 90449126 -->
unfiltered reads = `r sort_read_count` 
<!-- unfiltered reads = 163200129 -->

<!-- report number of snps we called nsnps (dat) and number of loci in IBS data (both in R) report number after filtering -->
number of snps = `r number_snps_after_filtering`
Number of snps= 71175?
number of IBS loci = `r nrow(IBS)`

##  Diversity Stats
<!-- see table X (embedding figures) -->
<!-- talk about any trends like this creek had higher diversity than this  -->
<!-- still need to work on mahanttan plot -> don't think we found anythign significant; there was not a position that greatly differed depending on the sample (in linux) -->
<!-- What are the thresholds at which the Ho values are significant. -->
The mean Ho values are relatively similar between the anadromous and residents individuals and between the CH and TK populations.  
<!-- Having trouble getting the fst values -> but should match up with PCA right?  -->
The fst values between the two populations and between anadromous and resident individuals are low, which indicated that there is not much genetic diversity between the two populations and two phenotypes. 
<!-- report stats in excel table  -->
There are also no private alleles between the two populations, which also indicated low genetic diveristy between the two populations of salmon. 

## Association 
<!--phenotypes and genetic basis -->
how to write up the results of the random forest? -> did not find anything significant because the actual results did not match the predicted. report r2 value 

## Influence of Anadromy on Population Structure
The distance from the other population's centroid is closer to the anadromous than the resident in one population. This indicates that the anadromous individuals within one population are more closely related to the individuals in the other population. 
<!-- Is there anything else I should add here?  -->
<!-- control shift c --> 
[@Gowen2014; @Ali2016]
Gowen et al.[-@Gowen2014]