---
title: "MASU Paper"
author: "William Hemstrom"
date: "11/21/2021"
output:
  word_document:
    reference_docx: "style_ref.docx"
  pdf_document:
    keep_tex: true
    fig_caption: true
    latex_engine: pdflatex
    fig_width: 7
    fig_height: 6
bibliography: "Citations.bib"
csl: "council-of-science-editors-author-date.csl"
header-includes:
  - \usepackage[left]{lineno}
  - \linenumbers
  - \usepackage{setspace}\doublespacing
  - \usepackage{sectsty}\sectionfont{\fontsize{12}{12}\selectfont}
  - \usepackage{sectsty}\subsectionfont{\normalfont\itshape\fontsize{12}{12}\selectfont}
  - \usepackage[round]{natbib}
indent: true
sansfont: Times New Roman
fontsize: 12pt
---

#Introduction 

<!-- Declines in migratory fishes -->
<!-- 	Trends and potential causes. -->

<!-- Importance of Masu -->
<!-- 	Vulnerability -->
<!-- 		Very prone to fishing pressure -->
<!-- 		Very limited distribution -->

<!-- Life History -->
<!-- 	What decides if an indivividual is a resident/anadromous? -->

<!-- Genetics of resident vs anadromous -->

<!-- What did we do and find? -->
<!-- 	Thesis -->
Due to the history of hydrologic modifications such as the development of reservoirs, dams, and irrigation systems, there has been an apparent decrease in the number of migratory fishes in the recent years. For example, on the Pacific coast there has been significant decreases in diversity within salmonid species due hydrologic and dam alterations. Atlantic salmon and anadromous species of sturgeons are also impacted negatively because of a loss of portions to their habitats upstream of dams. Overall, the diversity within migratory fish species have largely declined due to these dams blocking the fishes' migratory paths. One species of migratory fish that we will be discussing more in depth is the MASU salmon. 

MASU salmon, or the cherry salmon, have the most limited distribution among the Pacific salmon. Additionally, since they spend a lot of their life in freshwater rivers or streams, this makes them more inclined to be exposed to fishing and sportfishing pressures. Since they are more vulnerable than other salmon species, it is important to look into the details that influence their life cycles of being either resident or anadromous. Anadromous salmon are initially born in the rivers then migrate to the seas where they develop larger body sizes. These salmon eventually migrate back to the rivers in order to breed. The resident salmon in contrast sexually mature at a younger age and stay within the river systems throughout their entire lives (Morita and Nagasawa)

MASU salmon and their life cycles have been known to be primarily influenced by environmental factors such as temperature and other factors such as the salmon's body size, growth rate and lipid stores. Environmental factors such as climate change have been known to impact the decision of the salmon to be either resident or anadromous. It has been shown that the number of resident males have increased while the proportion of anadromous males and the proportion of delayed migrants within both females and males have decreased as the temperatures have increased. The increase in resident males due to increasing temperatures have led to a significant decrease in the number of anadromous individuals. Instead of looking into environmental influences, which have already beeen heavily studied, in this paper, we will be examining if there is a genetic basis within MASU salmon that will allow us to determine the salmon's life cycle. 

For comparison, within the O. mykiss species or rainbow trout, which is also a species of fish that exhibit either resident or anadromous life cycles, there is strong evidence suggesting a genetic basis on life history. The first evidence of a genetic basis for O. mykiss was when Neave (1944) conducted common garden and transplant experiments with sympatric rainbow trout and steelhead. After marking and releasing the progeny of the rainbow trout and steelhead, he found that life cycles were heritable after a larger proportion of rainbow trout offspring were recaptured over many years. Additionally, crosses between anadromous O. mykiss individuals resulted in the most smolts while crosses between resident individuals produced the fewest number of smolts. In a recent study by Hess et al. (2016), univariate analyses and a multivariate random forest (RF) machine learning algorithm was used to perform the association mapping of of 15 239 single nucleotide (SNPs) for adult migration-timing phenotype in steelhead. Using a univariate linear model resulted in the finding of three SNPs from a gene that was strongly linked with migration timing. Using multivariate analyses with RF resulted in the finding of 15 additional SNPs that were connected to 60% of the trait variation. 

Like how the rainbow trout and steelhead have a clear genetic basis for their life histories, we also wanted to investigate the genetic influences on MASU salmon and their life cycles using similar methods. We performed a random forest on the data, did association tests, and found diversity statistics, which did show any clear indicator of the genetic difference between the anadromous and resident MASU salmon. However, when creating a Principle Component Analysis chart (PCA) to look for patterns in relatedness between resident and anadromous individuals, we found that the anadromous individuals clustered significantly closer to the cluster center for individuals from the other population. Overall, from the DNA analysis methods used in this paper, there is not strong evidence that there is a genetic basis for MASU salmon life histories, but further studies using a bigger sample size would need to be conducted in order to investigate the influences of genetics on MASU salmon and their life cycles. 


```{r, echo=FALSE, include=FALSE}
dat<-read.table("../data/genotypes.geno")                        
meta<- read.table("../data/sample_meta.txt", header=TRUE)
snp.meta<- dat[,1:2]
colnames(snp.meta)<- c("chr", "position")
library(snpR)
dat<- import.snpR.data(dat[,-c(1:2)], snp.meta = snp.meta, sample.meta = meta)
dat<- filter_snps(dat, maf = 0.05, hwe = 0.000001)
number_snps_after_filtering <- nsnps(dat)

#calculating genetic diversity
dat <- calc_ho(dat, c("pop", "phen", "pop.phen"))
get.snpR.stats(dat, c("phen", "pop","pop.phen"), stats="ho") $weighted.means
get.snpR.stats(dat, c("pop", "phen", "pop.phen"))
dat <- calc_fis(dat, c("pop", "phen"))
dat <- calc_pi(dat, c("pop", "phen"))
get.snpR.stats(dat, c("pop"), stats="pi") $weighted.means
dat <- calc_private(dat, c("phen", "pop","pop.phen"))
get.snpR.stats(dat, c("pop"), stats="private") $weighted.means
get.snpR.stats(dat, c("pop", "phen", "pop.phen"))
dat <- calc_pairwise_fst(dat, c("phen", "pop","pop.phen"), boot = 100, boot_par = 2)
get.snpR.stats(dat, c("phen", "pop","pop.phen"), stats="fst") $weighted.means
get.snpR.stats(dat, c("pop", "phen"), stats= c("ho", "pi", "fst")) $weighted.means

gp<- run_genomic_prediction(dat, response = "phen", iterations = 10000, burn_in = 1000, thin = 100)
library(data.table)
cv<- cross_validate_genomic_prediction(dat, response = "phen", iterations = 10000, burn_in = 1000, thin = 100)

saveRDS(dat, "processed_genotypes.RDS")
saveRDS(gp, "gp.RDS")

```
# Methods:
## Genotyping 
We extracted DNA from our samples using the magnetic bead approach [@Ali2016]. We then prepared Restriction Associated Digest (RAD) libraries using the Pst1 restriction enzyme [@Ali2016] and sequenced the results on an Illumina Hi-Seq 4000 using 150bp paired-end sequencing. We aligned the data to the O. mykiss v6 reference genome [@Pearse2019] using Burrows-Wheeler Alignment algorithm Aligner [@Li2009]. We then filtered the data for PCR duplicates and improper pairs using SAMtools [@Li2009]. Next, we called the genotypes by using the SAMtools genotype likelihood models [@Li2009] used in the ANGSD software package with a minimum mapping and base call quality score of 20, a SNP p-value of 1e-8, and a minimum acceptable minor allele frequency of 0.05 [@Korneliussen2014]. We also used a minor allele frequency filter (MAF) and a Hardy-Weinberg equilibrium filter using snpR when calling the genotypes [@Wigginton2005]. 
<!-- flagstat google (data script i wrote) -->

```{r}
mat <- read.table("../angsdput.covMat")
IBS <- data.table::fread("../angsdput.covMat")
nrow(IBS)
IBS <- IBS[-32,-c(97, 32)]
pca_r <- eigen(IBS)
pca <- pca_r$vectors
write.table(pca[,1:20], "pca20.txt", col.names = FALSE, row.names = FALSE, quote=FALSE, sep="\t")
colnames(pca) <- paste0("PC", 1:ncol(pca))
meta <- read.table("../data/sample_meta.txt", header = TRUE)[-32,]
pca <- cbind(pca, meta)

loadings <- pca_r$values/sum(pca_r$values)
loadings <- round(loadings*100, 2)
pca$pop <- as.factor(pca$pop)
pca$phen <- as.factor(pca$phen)

library(ggplot2)
p1 <- ggplot(pca, aes(x = PC1, y = PC2, color = pop, shape = phen)) + theme_bw() +
  scale_color_viridis_d() + geom_point(size = 4) + 
  ggplot2::xlab(paste0("PC1 (", loadings[1], "%)")) + 
  ggplot2::ylab(paste0("PC2 (", loadings[2], "%)"))
p1
pca.centroids <- aggregate(pca[,1:2], list(stream = pca$pop), mean)
combine <- rbind(pca[, 1:2], pca.centroids[,-1])
dcombine<- dist(combine, upper = TRUE) 
dcombine<- as.matrix(dcombine)
dcombine<-dcombine[(nrow(dcombine) - nrow(pca.centroids) + 1):nrow(dcombine), -((nrow(dcombine) - nrow(pca.centroids) + 1):nrow(dcombine))]
dcombine <- t(dcombine) 
dcombine <- as.data.frame(dcombine)
dcombine$pop <- pca$pop
dcombine$phen <- pca$phen
colnames(dcombine)[1:2] <- c("CH", "TK")
dcombine$phen <- factor(dcombine$phen)
dcombine$pop <- factor(dcombine$pop)
dcombine$d <- 0
dcombine$d[dcombine$pop=="CH"] <- dcombine$TK[dcombine$pop=="CH"]
dcombine$d[dcombine$pop=="TK"] <- dcombine$CH[dcombine$pop=="TK"]
mod2 <- lm(d~phen, dcombine[-32,])
ggplot(dcombine, aes(x=pop, y=d, color=phen)) + geom_boxplot() + theme_bw() + scale_color_viridis_d(end = 0.5)

```

##  Diversity Stats
We calculated the observed heterozygosity (Ho), heterozygote/homozygote ratio per individual (Het/Hom), the average number of pairwise differences (pi), and fixation index (Fst) using the snpR R package [@Hohenlohe2010; @Wier1984; @Hemstrom2021]. 

```{r}
dat <- run_random_forest(dat, response = "phen", mtry = floor(nrow(dat)/2), num.trees = 100000)
RF <- dat$models
dat <- dat$data
summary(RF$.base_.base)

association <- data.table::fread("../association_out_cov.lrt0.gz")
association$p <- pchisq(association$LRT, 1, lower.tail = FALSE)
plot_manhattan(assocation, "p", chr = "Chromosome", bp= "Position")

```

## Association Testing
We performed a basic association test on the data using gmmat to correct for the related among individuals and population structure (). We then performed a random forest analysis in R-studio using the ranger R package [@Wright2017].  We used 100,000 trees and used mtry of half the number of snp's. We used a random forest to predict whether the samples would be resident or migratory. Using the BGLR program, we did a genomic prediction using BayesB distribution to estimate the heritability [@Meuwissen2001]. We used 1000 burn- ins and iterations=10,000. The heritability test was done to determine the degree in which phenotypic variation is due to genetic differences among the salmon populations. 


## Population Structure and Migratory Relatedness
In order to detect any underlying population structuring within our stream systems, we used the software package NGSadmix to group samples into *k* clusters for each value of *k* between one and seven [@Skotte693]. To do so, we first called genotype posterior probailites using ANGSD with the same parameters above in addition to (fill in here). To assess variance in clustering within a given *k* value, we ran NGSadmix ten times for each value of *k* and combined the resulting data using CLUMPP [@Jakobsson2007]. We used the pophelper and snpR packages to handle files and produce plots [@Hemstrom2021; @Francis2017].

For comparison, we also conducted a Principle Component Analysis chart (PCA) using the ANGSD-IBS method [@Korneliussen2014]. To look for patterns in relatedness between resident and andromous individuals, we calculated the mean PC1 and PC2 coordinates for each stream, then calculated the distance between each individual sample the mean for the other population.

<!-- include that we did a one-way anova to determine if either phenotype was more closely related to the mean of the opposite population -->


```{r ngsadmix,include=FALSE}
setwd("../ngsadmix/")
ngsadmix <- plot_structure(".qopt", paste0(meta$pop, "_", meta$phen), facet.order = c("CH_A", "CH_R", "TK_A", "TK_R"), k = 1:7)
ngsadmix <- ngsadmix$plot + xlab("Population/Phenotype")
```


<!-- Will need to fill in NGSadmix -->

# Results 
```{r}
MASU_data_table <- readr::read_delim("~/MASU_data_table", 
    delim = "\t", escape_double = FALSE, 
    trim_ws = TRUE, skip = 4)
MASU_data_table <- as.data.frame(MASU_data_table)
MASU_data_table[,3] <- gsub("%", "", MASU_data_table[,3])
MASU_data_table[,5] <- gsub("%", "", MASU_data_table[,5])
MASU_data_table[,3]<- as.numeric(MASU_data_table[,3])
MASU_data_table[,5]<- as.numeric(MASU_data_table[,5])
mean(MASU_data_table[,3], na.rm = TRUE)
mean(MASU_data_table[,5], na.rm = TRUE)
```

##Genotyping 
The number of unfiltered reads are `r sort_read_count`.  After filtering, the number of reads were `r sort_flt_read_count`. We called a total of `r number_snps_after_filtering` snps and the number of IBS loci were `r nrow(IBS)`. 
```{r setup, include=FALSE, eval=FALSE}
library(readr)
sort_read_count <- read.csv("~/Downloads/sort.read_count.txt", header = FALSE)
View(sort_read_count)
sort_read_count <- sum(sort_read_count$V1)

sort_flt_read_count <- read.csv("~/Downloads/sort.flt.read_count.txt", header = FALSE)
View(sort_flt_read_count)
sort_flt_read_count <- sum(sort_flt_read_count$V1)

```

flt read = `r sort_flt_read_count` 
<!-- flt read = 90449126 -->
unfiltered reads = `r sort_read_count` 
<!-- unfiltered reads = 163200129 -->
number of snps = `r number_snps_after_filtering`
Number of snps= 71175?
number of IBS loci = `r nrow(IBS)`

##  Diversity Stats
The mean Ho values are relatively similar between the anadromous and residents individuals and between the CH and TK populations. The fst values between the two populations and between anadromous and resident individuals are low, which indicated that there is not much genetic diversity between the two populations and two phenotypes. There are also no private alleles between the two populations, which also indicated low genetic diversity between the two populations of salmon. 

## Association 
The random forest predicted 25 out of 52 anadromous individuals correctly and 17 out of 44 resident individuals correctly for a total prediction accuracy of 38.6%. 

## Influence of Anadromy on Population Structure
We fit a simple linear regression with the distance to the other population cluster as the response variable and the migratory phenotype as the explanatory variable. We found that the anadromous individuals clustered signficiantly closer to the cluster center for individuals from the other population (one-way ANOVA, p-value = 0.0419). This indicates that the anadromous individuals within one population are more closely related to the individuals in the other population. 
<!-- The p value (phenR) is 0.03981775. The p value for the phenotypes is 0.0419.  -->


<!-- control shift c --> 
[@Gowen2014; @Ali2016]
Gowen et al.[-@Gowen2014]